### PR TITLE
Gui

### PR DESCRIPTION
**Pull request by collinsc**
# Description
The GUI/Utility methods for publishing status codes to the `com` channel. It provides a qt framework application that resides in the `tfr_mission_control` package. It also provides a utility class in `tfr_utilities` called `status_publisher` which publishes to the control center and the desired ros visibility level simultaneously.

So let's talk about the features of the gui, it starts in an unitialized state. At this point the user can start the mission autonomously, or start it in teleop, which makes the system ready to accept commands and initializes the clock. 

The teleop button is hidden in the top corner to make it harder to access, because that should only be used for testing.

Running autonomy will just flash a bunch of messages on the screen simulating the system running. You can switch to teleop mode by stopping autonomy. Switching in between teleop and autonomy will print some error messaages to stdout. This is caused by me calling the `isActive` methods on the action clients when they don't have a goal, to synchronize the servers. This error message is erroneous, and is a perfectly valid use case for synchronizing in between the two action clients. I have tested this thoroughly, and the forum community seems to agree this is not really an error, and sometimes a usefull thing to do.

So on to teleop. In telop mode we want to process user commands. In general some actions like digging and dumping we want to "fire and forget", but some like turntable movement, and directional control, we want to do until the user stops holding the button, then stop doing it.

This screen also accepts `WASD` control. Because of some details I go into in the documentation for the code I need to do manual debouncing of keyboard events. This means the controls might throw in an erroneous stop message if your keyboard repeated key OS delay is set longer than 0.3 seconds. So for a smooth experience change your repeated keys delay to something smaller in the keyboard settings menu. As shown here, you can look it up with unity dash.
screenshot from 2018-02-23 18-39-17

To test the scrolling window try the `toggle motors` button. It is the only thing hooked up there right now. Just click it a bunch of times.
# Testing
Go ahead and fire up the GUI, and try and break it, let me know of anything weird!
`roslaunch tfr_launch mission_control_test`

### CODE REVIEW

File: src/tfr_executive/src/teleop_action_server.cpp
Comment (RyanBerge) [inline src/tfr_executive/src/teleop_action_server.cpp]: Is there a reason this function isn't built on a switch statement?

File: src/tfr_executive/src/teleop_action_server.cpp
Comment (RyanBerge) [inline src/tfr_executive/src/teleop_action_server.cpp]: "Command Revieved"

File: src/tfr_executive/src/teleop_action_server.cpp
Comment (RyanBerge) [inline src/tfr_executive/src/teleop_action_server.cpp]: We return a `succeeded` state on unrecognized commands, even though we take no action.  Should probably set `failure` here and return.

**Review by RyanBerge — CHANGES_REQUESTED** [review]
Just a couple questions and comments.  One more, which I can't put in an inline comment because the area isn't part of the diff for this PR, is that the `frequency` value in `teleop_action_server` is not currently being used.  You should do one of the following:

- Remove it
- Add preemption checks
- Add TODO comments where preemption checks should be

I didn't look through the QT code very closely because I'm not familiar with the API. If there's something seriously wrong with it, we should notice during testing.

File: src/tfr_executive/src/teleop_action_server.cpp
Comment (collinsc) [reply src/tfr_executive/src/teleop_action_server.cpp]: fixed, but we will be removing these on a future release, so they don't need to be too pretty

File: src/tfr_executive/src/teleop_action_server.cpp
Comment (collinsc) [reply src/tfr_executive/src/teleop_action_server.cpp]: It was a while ago, but I think that was a relic of a typing logic error with enum classes. I put it in an else if telling my self "put it in a switch if you ever figure that out" then I figured it out and forgot to the logic it in the switch.

File: src/tfr_executive/src/teleop_action_server.cpp
Comment (collinsc) [reply src/tfr_executive/src/teleop_action_server.cpp]: The state we want to signal that behavior is called `aborted` but I know what you mean, sure

**Comment by collinsc** [comment]
@RyanBerge we will use that frequency value during digging/dumping, and that is a task for systems integration. 

Looking through my comments, it looks like I put a TODO there but it wasn't a very specific comment, I made it longer and made sure to mention that was the place we need to check for preemption

File: src/tfr_executive/src/teleop_action_server.cpp
Comment (AdamuKaapan) [inline src/tfr_executive/src/teleop_action_server.cpp:171]: Are we operating the digging as single digging sets (scoops) or just saying "keep digging until I tell you to stop"?

File: src/tfr_mission_control/plugin.xml
Comment (AdamuKaapan) [inline src/tfr_mission_control/plugin.xml:10]: We should probably fill these out: even though we probably won't ever see them, just for good practice (although no big deal if it doesn't get changed, so don't let this block the PR).

File: src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp
Comment (AdamuKaapan) [inline src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp:62]: I'm assuming that "Strong Focus" means that keyboard events outside the window don't trigger events here, correct?

Comment (AdamuKaapan) [reply src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp:62]: If that's the case that's a really good thing

File: src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp
Comment (AdamuKaapan) [inline src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp:216]: The naming here is really confusing, just because I'm seeing "set autonomy" maps to "enable the teleop button", which feels really backwards to me. It just worries me that something's gonna get flipped, so I'd like to see a bit more documentation/a name change to make this clearer.

File: src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp
Comment (AdamuKaapan) [inline src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp:317]: Just 'cause I'm curious, is this some sort of special Qt preprocessor thing?

File: src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp
Comment (AdamuKaapan) [inline src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp:366]: What's the reasoning for having softwareStop() so late in this method (vs. goAutonomousMode() where it's the first line)? I'm guessing it's for letting autonomy hand over control of the actuators first before we stop them?

**Review by AdamuKaapan — CHANGES_REQUESTED** [review]
Nothing really breaking here, just had a few questions/little comments that I think should at least be considered before I actually approve it.

File: src/tfr_executive/src/teleop_action_server.cpp
Comment (collinsc) [reply src/tfr_executive/src/teleop_action_server.cpp:171]: I give it the time allowed to dig just like the specification

File: src/tfr_mission_control/plugin.xml
Comment (collinsc) [reply src/tfr_mission_control/plugin.xml:10]: sure I don't pretend to be a gui expert, in fact quite the opposite. I just wanted something that turned on.

File: src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp
Comment (collinsc) [reply src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp:62]: I also don't know I just followed the tutorial to get it to turn on, not a front end dev.

File: src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp
Comment (collinsc) [reply src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp:216]: it's a window state, I think it makes sense. The only button you should be able to press when you are in autonomous mode is the enable teleop.

File: src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp
Comment (collinsc) [reply src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp:317]: It's sends the string to the next event in the chain of events to process

File: src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp
Comment (collinsc) [reply src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp:366]: It's just an extra measure. Notice how we are letting the autonomous action server preempt itself? on line 364. I just did it there in case anything failed with the preemption. It doesn't hurt us.

**Comment by collinsc** [comment]
@AdamuKaapan I answered your questsions

File: src/tfr_executive/src/teleop_action_server.cpp
Comment (RyanBerge) [reply src/tfr_executive/src/teleop_action_server.cpp:171]: @AdamuKaapan Executive should just know how to tell it to go with a timer. The Mining System is in charge of deciding how many scoops that is. This looks correct to me.

**Review by RyanBerge — APPROVED** [review]

File: src/tfr_executive/src/teleop_action_server.cpp
Comment (AdamuKaapan) [reply src/tfr_executive/src/teleop_action_server.cpp:171]: Yeah, that makes sense; I just wasn't sure if we wanted to do it a bit more granularly because it's manual operation, but I'm totally fine with this too.

File: src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp
Comment (AdamuKaapan) [reply src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp:317]: Yeah, I was just wondering about the `emit emitStatus(msg);`, that seems like preprocessor stuff unless there's some weird `emit` keyword in C++ I don't know about, hahaha.

File: src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp
Comment (AdamuKaapan) [reply src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp:366]: Makes sense, yeah, if everything else also stops the motor.

**Review by AdamuKaapan — APPROVED** [review]

**Pull request merged**
